### DISCUSSION TITLE
New-user: When is inference finished? Prediction vs #-annotations performance?

### DISCUSSION THREAD

**Discussion by aamouselab** [talmolab/sleap#1209] [Help!] [answered]
Hi Talmo lab, 

I'm trying sleap for the first time, on a series of videos of single animals, with limited coding skills. I have just reached the inference step. Every time sleap reaches the last predicted frame, it somehow restarts the prediction. Is this behavior expected and I will find out at some point that it is done with predictions or is there an issue and should I restart the inference step? 

Thank you very much for your help! 

[IMG_0584 2.mov.zip](https://github.com/talmolab/sleap/files/10775314/IMG_0584.2.mov.zip)

**Comment by roomrys** [comment]
Hi @aamouselab,

This behavior is expected when predicting on multiple videos, although, from your video it seemed that you were predicting on the same number of frames each time (14,986), so I couldn't verify this. 

Some follow-up questions: Did you predict on multiple videos? Do your videos contain the same number of frames (14,986)? Did the predictions workout? Is there anything posted in your terminal and/or any dialogs?

Thanks,
Liezl

**Comment by aamouselab** [comment]
Hi Liezl, 

Thank you for getting back to me. Yes, I have the same number of frames all the time. I do not remember exactly what was on the terminal last time because it was running indefinitely with no way of finding out whether it was working or not so I stopped it and tried exporting data to get an idea of whether something had worked or not. I obtained a bunch of h5 files at the end, which let me think that it did work. I did this first test by following the tutorial but it looked like training on 20 frames was not enough.

I thus duplicated the project, trained it on 361 frames this time and then ran inference on the whole videos. I got the following message: "Inference ran on 361 frames. Instances were predicted on 349 frames (12 frames with no instances found). No message on the terminal. I assumed it was done and exported the data. I am still very new to this and used the jupyter notebook provided in the tutorial to take a look at my data. The graphs indicating nodes location are identical to the ones obtained with training on 20 frames. It is hard to believe that no improvement would be obtained so I'm thinking that something went wrong during the training and or inference steps... The metrics do show less pixel error for some of the nodes that I tracked. I have a hard time understanding when inferences are actually done and why the h5 data files are not different after creating a new project with training on +300 frames....

**Reply by roomrys** [reply, marked as answer]
Hi @aamouselab,

Inference is run on a per-video basis. Hence, for each video that you are running inference on, a progress bar will pop-up indicating the progress made for that video - _perhaps we should indicate which video is being predicted on via either the progress bar or a message in the terminal_. Likewise, when exporting SLEAP analysis files, each video in the project will have it's own analysis file.

If you have not gotten improvement from labeling more frames, then it could be an issue with the model hyperparameters - do you mind sharing a screenshot of your model configurations (Training Pipeline), including the blue box (Receptive Field) overlaid on the image.

#### Side-note: Workflow tips for new-users

As a [general workflow](https://sleap.ai/tutorials/initial-labeling.html?highlight=labeling%20suggestions), we suggest generating a subset of frames to label (Labeling Suggestions) and labeling a fraction of those frames - leaving a few to run predictions on for [prediction-assisted labeling](https://sleap.ai/tutorials/assisted-labeling.html). You can always add more frames to the Labeling Suggestions as needed. 

Sticking to labeling and predicting on this subset will make the training/prediction much faster and more organized. We do not recommend predicting on the entire video(s) while doing the human-in-the-loop training. You should only predict on the entire video when predictions look good and you are ready to perform tracking (or if you are curious!).

Thanks,
Liezl
